dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/relationship/hibernate/HibernateRelationshipStore.java

```diff
@@ -75,36 +83,117 @@ public HibernateRelationshipStore( SessionFactory sessionFactory, JdbcTemplate j
     {
-        CriteriaBuilder builder = getCriteriaBuilder();
+        TypedQuery<Relationship> relationshipTypedQuery = getRelationshipTypedQuery( pi,
+            pagingAndSortingCriteriaAdapter );
 
-        return getList( builder, newJpaParameters( pagingAndSortingCriteriaAdapter, builder )
-            .addPredicate( root -> builder.or(
-                builder.equal( root.join( "from" ).get( "programInstance" ), pi ),
-                builder.equal( root.join( "to" ).get( "programInstance" ), pi ) ) ) );
+        return getList( relationshipTypedQuery );
     }
 
     @Override
     public List<Relationship> getByProgramStageInstance( ProgramStageInstance psi,
         PagingAndSortingCriteriaAdapter pagingAndSortingCriteriaAdapter )
+    {
+        TypedQuery<Relationship> relationshipTypedQuery = getRelationshipTypedQuery( psi,
+            pagingAndSortingCriteriaAdapter );
+
+        return getList( relationshipTypedQuery );
+    }
+
+    private <T extends IdentifiableObject> TypedQuery<Relationship> getRelationshipTypedQuery( T entity,
+        PagingAndSortingCriteriaAdapter pagingAndSortingCriteriaAdapter )
     {
         CriteriaBuilder builder = getCriteriaBuilder();
 
-        return getList( builder, newJpaParameters( pagingAndSortingCriteriaAdapter, builder )
-            .addPredicate( root -> builder.or(
-                builder.equal( root.join( "from" ).get( "programStageInstance" ), psi ),
-                builder.equal( root.join( "to" ).get( "programStageInstance" ), psi ) ) ) );
+        CriteriaQuery<Relationship> relationshipItemCriteriaQuery = builder.createQuery( Relationship.class );
+        Root<Relationship> root = relationshipItemCriteriaQuery.from( Relationship.class );
+
+        setRelationshipItemCriteriaQueryExistsCondition( entity, builder, relationshipItemCriteriaQuery, root );
+
+        return getRelationshipTypedQuery( pagingAndSortingCriteriaAdapter,
+            builder, relationshipItemCriteriaQuery, root );
+    }
+
+    private <T extends IdentifiableObject> void setRelationshipItemCriteriaQueryExistsCondition( T entity,
+        CriteriaBuilder builder,
+        CriteriaQuery<Relationship> relationshipItemCriteriaQuery, Root<Relationship> root )
+    {
+        String relationShipEntityType = getRelationShipEntityType( entity );
+
+        Subquery<RelationshipItem> relationshipItemFromSubQuery = relationshipItemCriteriaQuery
+            .subquery( RelationshipItem.class );
+        Root<RelationshipItem> relationshipItemFromRoot = relationshipItemFromSubQuery.from( RelationshipItem.class );
+
+        Predicate predicateFromEqualsId = builder.equal( root.get( "from" ), relationshipItemFromRoot.get( "id" ) );
+        Predicate predicateFromEqualsTei = builder.equal( relationshipItemFromRoot.get( relationShipEntityType ),
+            entity.getId() );
+
+        relationshipItemFromSubQuery.where( predicateFromEqualsId, predicateFromEqualsTei );
```

Comment:
I would inline variables like this one that are only ever used once. Unless it makes the code less readable because its such a huge expression. As a reader I don't want to read, have to remember predicateFromEqualsId only to have to remove it from my mental context since it only lives for 1 or 2 lines of code.